File: graphs/200_number_of_islands/200_number_of_islands.py
"""
LeetCode 200. Number of Islands
Link: https://leetcode.com/problems/number-of-islands/
Difficulty: Medium
Primary: Graph
Tags: [graph, bfs, dfs, matrix, connected-components]
Solved: 2025-11-14
"""

from typing import List

class Solution:

    # --- Approach 1: Depth First Search ---
    # Time: O(m * n), Space: O(m * n) 
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])

        def dfs(r, c):
            if r < 0 or c < 0 or r == ROWS or c == COLS or grid[r][c] == '0':
                return 
                
            grid[r][c] = '0' 

            dfs(r + 1, c)
            dfs(r, c - 1)
            dfs(r - 1, c)
            dfs(r, c + 1)

        res = 0
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == '1':
                    dfs(r, c)    
                    res += 1

        return res
    

# --- Example Test ---
if __name__ == "__main__":
    solution = Solution()

File: graphs/200_number_of_islands/200_number_of_islands.java
/*
LeetCode 200. Number of Islands
Link: https://leetcode.com/problems/number-of-islands/
Difficulty: Medium
Primary: Graph
Tags: [graph, bfs, dfs, matrix, connected-components]
Solved: 2025-11-14
*/

class Solution {

    // --- Approach 1: Depth First Search ---
    // Time: O(m * n), Space: O(m * n) 
    private int ROWS;
    private int COLS;

    public int numIslands(char[][] grid) {
        ROWS = grid.length;
        COLS = grid[0].length;
        int count = 0;

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (grid[r][c] == '1') {
                    dfs(r, c, grid);
                    count++;
                }
            }
        }
        return count;
    }

    private void dfs(int row, int col, char[][] grid) {
        if (row < 0 || row == ROWS || col < 0 || col == COLS || grid[row][col] == '0') {
            return;
        }

        grid[row][col] = '0';
        dfs(row, col + 1, grid);
        dfs(row, col - 1, grid);
        dfs(row + 1, col, grid);
        dfs(row - 1, col, grid);
    }




    // --- Example Test ---
    public static void main(String[] args) {
        Solution sol = new Solution();

        char[][] grid1 = {
            {'1','1','1','1','0'},
            {'1','1','0','1','0'},
            {'1','1','0','0','0'},
            {'0','0','0','0','0'}
        };
        System.out.println(sol.numIslands(grid1)); // Expected: 1
    }
    
}